### DISCUSSION TITLE
Bind to device *and IP*

### DISCUSSION THREAD

**Discussion by orgads** [curl/curl#13688] [Ideas]
If a device has several IP addresses, it should be possible to force one of them.

We have a patch that implements this as an additional option in libcurl (and a command-line argument).

While preparing it for PR, I was wondering if it would be better to extend the existing `if!`/`host!` notation, and do something like `if!eth0!10.2.3.4` instead.

Current implementation (in our patched curl):
```c
    curl_easy_setopt(curl, CURLOPT_INTERFACE, "eth0"); /* or "if!eth0" */
    curl_easy_setopt(curl, CURLOPT_INTERFACE_IP, "10.2.3.4");
```

Proposed change:
```c
    curl_easy_setopt(curl, CURLOPT_INTERFACE, "if!eth0!10.2.3.4");
```

If the second option is preferred, then I can also drop the extra command-line argument.

Will you accept such a feature? If you will, which of these options do you prefer?

**Comment by orgads** [comment]
@bagder @vszakats ?

**Reply by vszakats** [reply]
With the single option, perhaps a distinct new prefix, e.g. `ifhost!eth1!10.2.3.4`, would make it more unambiguous to write/read and easier to parse?

Regardless of the prefix chosen, when passed to older curl without support for this, it will result in an unexpected failure, while with the extra `CURLOPT` option the additional IP/host constraint would just be a no-op. Though perhaps a failure is the better outcome here.

As for `CURLOPT_INTERFACE_IP`, assuming it also support hosts, would  `CURLOPT_INTERFACE_HOST` be a better name? (rhyming with the `host!` prefix). The downside of the extra option is that some might assume this also works standalone to bind to an IP.

I tend to prefer the the single option with a distinct new prefix.

**Reply by orgads** [reply]
I like `ifhost!`. @bagder do you have a preference?

**Comment by bagder** [comment]
How exactly does this differ from binding to the IP address alone? Surely a host cannot have the same IP on multiple network interfaces?

**Reply by orgads** [reply]
1. See #13613 for an example why IP address is not enough.
2. You'll be surprised, but we do have customers with the same IP address on different network interfaces, don't ask 😅

**Reply by bagder** [reply]
I don't think #13613 explained it, hence me asking. And now I am wiser! :smile:

**Reply by orgads** [reply]
These are 2 different reasons. Regarding #13613, even if you don't have the same IP on different devices, the kernel can route outgoing packets from a different device, not necessarily the one with the bound address.

**Comment by orgads** [comment]
Draft PR here: https://github.com/curl/curl/pull/13719

There are 2 commits there. The first one adds an option in the library, and the second one implements `ifhost!` (and reverts most of the first one).

Once you state your preference, I'll either squash them or leave only the first one.

**Comment by orgads** [comment]
Ok, I implemented `ifhost!` notation in #13719. Closing this discussion.
